torch/cuda/jit.py

```diff
@@ -0,0 +1,156 @@
+        else:
+            raise ValueError('Expected a torch.device with a specified index '
+                             'or an integer, but got:{}'.format(device))
+    device_index = -1
+    if isinstance(device, str):
+        device = torch.device(device)
+
+    if isinstance(device, torch.device):
+        if not allow_cpu and device.type == 'cpu':
+            raise ValueError('Expected a non cpu device, but got: {}'.format(device))
+        device_index = -1 if device.type == 'cpu' else torch.cuda._cuda_getDeviceIndex(device)
+
+    if isinstance(device, int):
+        device_index = device
+
+    return device_index
+
+class device(object):
+    r"""Context-manager that changes the selected device.
+    This is similar to device (torch.device or int), but has been
+    introduced for jit compatibility
+
+    Arguments:
+        device (torch.device or int): device index to select. It's a no-op if
+            this argument is a negative integer or ``None``.
+    """
+
+    def __init__(self, device: _device):
+        self.idx = get_device_index(device)
+        self.prev_idx = -1
+
+    def __enter__(self):
+        if self.idx == -1:
+            return
+        self.prev_idx = torch.cuda.current_device()
+
+        if self.prev_idx != self.idx:
+            torch.cuda._cuda_setDevice(self.idx)
+
+    def __exit__(self, type: Any, value: Any, traceback: Any):
+        if self.prev_idx != self.idx:
+            torch.cuda._cuda_setDevice(self.prev_idx)
+
+class StreamContext(object):
+    r"""Context-manager that selects a given stream.
+
+    All CUDA kernels queued within its context will be enqueued on a selected
+    stream.
+
+    Arguments:
+        StreamContext (Stream): selected stream. This manager is a no-op if it's
+            ``None``.
+
+    .. note:: Streams are per-device. If the selected stream is not on the
+        current device, this function will also change the current device to
+        match the stream.
+    """
+
+    def __init__(self, stream: 'torch.classes.cuda.Stream'):
+        self.idx = get_device_index(device=None, optional=True)
+        self.cur_stream = stream
+        self.src_prev_stream = stream
+        self.dst_prev_stream = stream
+
+    def __enter__(self):
+        if self.idx == -1:
+            return
+        self.src_prev_stream = torch.cuda.current_stream(self.idx)
+        if self.idx != self.cur_stream.device_index():
+            with device(self.cur_stream.device()):
+                self.dst_prev_stream = torch.cuda.current_stream(self.idx)
+            torch.cuda._cuda_setDevice(self.cur_stream.device_index())
+        torch.cuda._cuda_setStream(self.cur_stream)
+
+    def __exit__(self, type: Any, value: Any, traceback: Any):
+        if self.src_prev_stream.device_index() != self.cur_stream.device_index():
+            torch.cuda._cuda_setStream(self.dst_prev_stream)
+            torch.cuda._cuda_setDevice(self.idx)
+        torch.cuda._cuda_setStream(self.src_prev_stream)
```

Comment:
Should this line be in an `else` block? If the device (device A, let's say) of the stream we switched to is different from the active device (device B) in the scope of this context manager, the active stream of device B was never changed, so we don't need to change it back.

Reply:
Similar to previous comment